File: cpp_solutions/131.cpp
class Solution {
public:
    vector<vector<string>> res;
    void backtracking(int index, vector<string>& path, string& s, vector<vector<bool>>& dp) {
        if (index == s.size()) {
            res.push_back(path);
            return;
        }
        for (int i = index; i < s.size(); ++i) {
            if (dp[index][i]) {
                path.push_back(s.substr(index, i - index + 1));
                backtracking(i + 1, path, s, dp);
                path.pop_back();
            }
        }
    }
    vector<vector<string>> partition(string s) {
        int n = s.size();
        vector<vector<bool>> dp(n, vector<bool>(n, false)); // s[i : j] is palindrome or not
        for (int i = 0; i < n; ++i) dp[i][i] = true;
        for (int length = 2; length <= n; ++length) {
            for (int i = 0; i < n; ++i) {
                int j = i + length - 1;
                if (j > n) break;
                if (length == 2) {
                    dp[i][j] = (s[i] == s[j]);
                }
                else {
                    dp[i][j] = (dp[i + 1][j - 1] && (s[i] == s[j]));
                }
            }
        }

        vector<string> path;
        backtracking(0, path, s, dp);
        return res;
    }
};

File: rust_solutions/131.rs
impl Solution {
    pub fn partition(s: String) -> Vec<Vec<String>> {
        let n: usize = s.len();
        let mut dp: Vec<Vec<bool>> = vec![vec![false; n]; n];
        let chs: Vec<char> = s.chars().collect();
        for right in 0..n {
            for left in 0..=right {
                if chs[left] == chs[right] && (right - left <= 2 || dp[left + 1][right - 1]) {
                    dp[left][right] = true;
                }
            }
        }

        let mut path: Vec<String> = Vec::new();
        let mut paths: Vec<Vec<String>> = Vec::new();
        Self::backtracking(0, &mut path, &s, &dp, &mut paths);
        return paths;
    }
    fn backtracking(
        index: usize,
        path: &mut Vec<String>,
        s: &str,
        dp: &Vec<Vec<bool>>,
        paths: &mut Vec<Vec<String>>,
    ) {
        let n: usize = s.len();
        if index == n {
            paths.push(path.to_vec());
            return;
        }
        for i in index..n {
            if dp[index][i] {
                path.push(s[index..1 + i].to_owned());
                Self::backtracking(1 + i, path, s, dp, paths);
                path.pop();
            }
        }
    }
}
